Automattic/newspack-listings
# feat(ugc): self-serve listings purchase block

### All Submissions:






### Changes proposed in this Pull Request:



Implements a purchase form block for self-serve listings. This PR only implements the block in the editor and its front-end output, not the handling of the form submission (which will be in a separate PR). The look and behavior of the block are heavily inspired by the Donate block in Newspack Blocks, both for ease of feature development and to maintain a consistent UX across Newspack products.

The form is designed to contain as few fields as possible in order to reduce purchase friction. I'm hoping that the actual content editing of purchased listings can happen directly in the block editor, but more to come on that in a future PR. I may add/remove/change the inputs as needed as we develop the full purchase flow.

@thomasguillot and/or @laurelfulford, feel free to suggest design-related changes or just push your own changes as needed.

As with all purchase-related self-serve features, this block is gated behind the `NEWSPACK_LISTINGS_SELF_SERVE_ENABLED` environment flag, so if this flag isn't set to `true` in `wp-config.php`, it won't be available.

<img width="785" alt="Screen Shot 2021-09-24 at 3 30 09 PM" src="https://user-images.githubusercontent.com/2230142/134741843-e4ca74ac-3752-4cbe-a665-12a5afbc4756.png">

### How to test the changes in this Pull Request:

1. In `wp-config.php`, set `define( 'NEWSPACK_LISTINGS_SELF_SERVE_ENABLED', true );` to enable the feature.
2. In a post or page, confirm there's a new block in the block inserter under the "Newspack" category:

<img width="340" alt="Screen Shot 2021-09-24 at 3 33 34 PM" src="https://user-images.githubusercontent.com/2230142/134742026-de37f634-d12c-4ab4-b1b8-09487516c412.png">

3. Insert the block into your content. Note that in the editor, the form is interactive but you can't submit it. You can edit certain copy within the block preview, such as the descriptions for single and subscription products, and the Submit button text. Play with the the block attributes and see how the different attribute settings behave on the front-end. Make sure to preview at various viewports to ensure that the styles look good in all cases.
4. Confirm that when submitting the form on the front-end, it doesn't do anything at the moment, but it does append URL params according to the form values you enter (these will be intercepted and used to add the corresponding products to the WC cart before redirecting to Checkout, just like with the Donate flow). The URL should look something like:

```
<PAGE URL>/?listing-purchase-type=single&listing-single-type=classified&listing-featured-upgrade=on&listing-title=My+Self-Serve+Listing
```
5. In `wp-config.php`, set `define( 'NEWSPACK_LISTINGS_SELF_SERVE_ENABLED', false );` or delete the constant entirely to disable the feature.
6. Confirm that in the editor, the block is no longer insertable (and the block you previously inserted shows a "not supported" error). Confirm on the front-end that the block renders no markup.
7. For kicks, you can re-enable the feature and activate a non-Newspack theme to confirm that the styles look surprisingly intact even without the Newspack CSS:

<img width="629" alt="Screen Shot 2021-09-24 at 3 42 21 PM" src="https://user-images.githubusercontent.com/2230142/134742741-dec76c87-f03e-4711-bb88-d306bae4e7f4.png">

### Other information:
